### PR TITLE
Records page UI refresh

### LINKED ISSUE #478: Records Page UI Refresh
Slight adjustments to the Records Page UI ([Figma](https://www.figma.com/file/brkAAatBCcvSnLJOIxZJIM/recidiviz-x-allison?node-id=551-23482&t=KCQ5xrxOzqse46qL-0) for full designs). This consists of a few changes:
* Larger tabs for All Reports, Draft, etc.
* New, larger buttons to replace the text-only "Select" and "New" buttons.
     * "New" should change to "New Report"
     * "Select" should be replaced with a new "Bulk Actions" menu
     * Right now, the only item in this menu should be "Delete". We will add "Publish" and "Unpublish" later
* When a bulk action is selected:
     * The rows should be come selectable
     * There should be new buttons: One that says "Cancel" and one that "Delete" (the "Delete" button only appears if the user has selected some rows)
     * The "All Reports", "Draft", etc tabs should be replaced with copy that says "Select record(s) to delete ..." and "X records selected"
     * When the user clicks "Delete", show a [confirmation modal](https://www.figma.com/file/brkAAatBCcvSnLJOIxZJIM/recidiviz-x-allison?node-id=551-23984&t=U6CdnjLvAoM06qOP-0)

### PR DESCRIPTION
**Pull request by ivan-kishko**
## Description of the change

> Changes to records page UI

## Type of change
Non-breaking refactor

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #478

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [x] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: publisher/src/pages/Reports.tsx [hunk 1]
```diff
@@ -317,30 +336,74 @@ const Reports: React.FC = () => {
...
+                    selectedRecords.length > 1 ? "s" : ""
+                  } selected`
+                : "Select draft(s) to publish..."}
+            </BulkActionModeTitle>
+          )}
+
+          {bulkAction === "unpublish" && (
+            <BulkActionModeTitle>
+              {selectedRecords.length
+                ? `${selectedRecords.length} record${
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:392]: same note on using records constant here & L395

**Review by mxosman — COMMENTED** [review]
Looks and feels great to me, Ivan! One potential enhancement is to prevent the user from entering a bulk action mode/displaying some message if there are no records to act on. 

For example - when you have no records and click on Bulk Actions and choose 'Publish' | 'Unpublish' | 'Delete' -- where it says 'Select draft(s) to publish...' could instead say something like "There are no records to ___" -- or we prevent the user from choosing any of those options and display "There are no records to ___ (publish, unpublish, delete)".

File: publisher/src/pages/Reports.tsx [hunk 2]
```diff
@@ -349,47 +412,70 @@ const Reports: React.FC = () => {
...
+                        textColor="blue"
+                        onClick={() => navigate("create")}
+                      >
+                        + New Record
+                      </ReportActionsButton>
                     </>
                   )}
 
-                  {selectionMode && (
+                  {!!bulkAction && (
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:461]: Question -- any benefit to `!!bulkAction` vs just using `bulkAction` since it'll be (string | undefined)?

Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx:461]: not really, just explicitly make this a boolean)

File: publisher/src/components/Reports/RemoveRecordsModal.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+export const RemoveRecordsModal: React.FC<RemoveRecordsModalProps> = ({
+  selectedRecords,
+  closeModal,
+  confirmRemoveRecords,
+}) => {
+  return (
+    <RemoveRecordsModalWrapper>
+      <RemoveRecordsModalContainer>
+        <RemoveRecordsModalTitle>
+          Delete <span>{selectedRecords}</span> record
```
[the remark this answers is not in the document]
Comment (ivan-kishko) [reply publisher/src/components/Reports/RemoveRecordsModal.tsx]: [c54aae4](https://github.com/Recidiviz/justice-counts/pull/488/commits/c54aae418a4bd5cdd6baa1de35fd37293a0ee841)

File: publisher/src/pages/Reports.tsx [hunk 4]
```diff
@@ -107,14 +115,19 @@ const Reports: React.FC = () => {
...
+  const [selectedRecords, setSelectedRecords] = useState<number[]>([]);
+  const [bulkAction, setBulkAction] = useState<RecordsBulkAction | undefined>(
+    undefined
+  );
+  const [isRemoveRecordsModalOpen, setIsRemoveRecordsModalOpen] =
+    useState(false);
+
+  const activateBulkAction = (action: RecordsBulkAction) =>
+    setBulkAction(action);
+  const deactivateBulkAction = () => setBulkAction(undefined);
```
[the remark this answers is not in the document]
Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx]: [c54aae4](https://github.com/Recidiviz/justice-counts/pull/488/commits/c54aae418a4bd5cdd6baa1de35fd37293a0ee841)

File: publisher/src/pages/Reports.tsx [hunk 5]
```diff
@@ -107,14 +115,19 @@ const Reports: React.FC = () => {
...
-  const addOrRemoveReportToDelete = (reportID: number) =>
-    setReportsToDelete((prev) =>
+  const [selectedRecords, setSelectedRecords] = useState<number[]>([]);
+  const [bulkAction, setBulkAction] = useState<RecordsBulkAction | undefined>(
+    undefined
+  );
+  const [isRemoveRecordsModalOpen, setIsRemoveRecordsModalOpen] =
+    useState(false);
+
+  const activateBulkAction = (action: RecordsBulkAction) =>
```
[the remark this answers is not in the document]
Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx]: [c54aae4](https://github.com/Recidiviz/justice-counts/pull/488/commits/c54aae418a4bd5cdd6baa1de35fd37293a0ee841)

File: publisher/src/pages/Reports.tsx (hunk 1 above)
Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx]: [c54aae4](https://github.com/Recidiviz/justice-counts/pull/488/commits/c54aae418a4bd5cdd6baa1de35fd37293a0ee841)

File: publisher/src/pages/Reports.tsx (hunk 2 above)
Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx]: [c54aae4](https://github.com/Recidiviz/justice-counts/pull/488/commits/c54aae418a4bd5cdd6baa1de35fd37293a0ee841)

**Comment by ivan-kishko** [comment]
> Looks and feels great to me, Ivan! One potential enhancement is to prevent the user from entering a bulk action mode/displaying some message if there are no records to act on.
> 
> For example - when you have no records and click on Bulk Actions and choose 'Publish' | 'Unpublish' | 'Delete' -- where it says 'Select draft(s) to publish...' could instead say something like "There are no records to ___" -- or we prevent the user from choosing any of those options and display "There are no records to ___ (publish, unpublish, delete)".

Hello @mxosman !) So about this, I made bulk action disabled if there are no records. Also I made "unpublish" disabled if draft tab selected and same for published tab and "publish" action. 1 question though to you and maybe @lilidworkin and @hobuobi. On "All records" tab you can without any restrictions choose either "publish" or "unpublish" action. I guess I have to do something with records on this tab like if you choose "publish" action disable all published records for selection or something like that, I think you get the idea, LMK what do you think?

**Comment by lilidworkin** [comment]
This looks awesome @ivan-kishko !

> So about this, I made bulk action disabled if there are no records. 

Perfect, that looks great to me!

> Also I made "unpublish" disabled if draft tab selected and same for published tab and "publish" action (I could make this menu items disappear as it pointed out in figma notes but then I had to return <></> in conditional statement since DropdownMenu cannot take null/false as a child so I decided that disabled is cleaner).

Great solution, I think disabling those actions looks good!

> One question though to you and maybe @lilidworkin and @hobuobi. On "All records" tab you can without any restrictions choose either "publish" or "unpublish" action. I guess I have to do something with records on this tab like if you choose "publish" action disable all published records for selection or something like that, I think you get the idea, LMK what do you think?

Great question! Yea, is it possible to disable selection of published records for "publish" and likewise, disable selection of draft records for "unpublish"? I would do whatever is easiest here.

Also @ivan-kishko / @mxosman , I haven't tested this so it might work correctly, but a few things to keep in mind re: permissions:
* JC ADMIN and AGENCY ADMIN should be able to create records, but not CONTRIBUTOR
* _Only_ JC ADMIN can delete records
* Eventually, everyone can publish/unpublish. But since this functionality is not implemented yet, can we hide it for everyone except JC ADMINS? I think this means that the entire Bulk Actions menu would be hidden for everyone except JC ADMINS right now.

File: publisher/src/pages/Reports.tsx (hunk 2 above)
Comment (mxosman) [reply publisher/src/pages/Reports.tsx]: Thanks, Ivan! I have no preference - was mostly curious.

**Comment by mxosman** [comment]
> Hello @mxosman !) So about this, I made bulk action disabled if there are no records. Also I made "unpublish" disabled if draft tab selected and same for published tab and "publish" action (I could make this menu items disappear as it pointed out in figma notes but then I had to return `<></>` in conditional statement since DropdownMenu cannot take null/false as a child so I decided that disabled is cleaner).

Hey Ivan! Nice touch - I like that option! Makes sense regarding the dropdown - I agree it is cleaner.

> One question though to you and maybe @lilidworkin and @hobuobi. On "All records" tab you can without any restrictions choose either "publish" or "unpublish" action. I guess I have to do something with records on this tab like if you choose "publish" action disable all published records for selection or something like that, I think you get the idea, LMK what do you think?

Yeah - this will be an interesting challenge to orchestrate. I do like the idea of adjusting the records based on which bulk action you select - like your example - where if you have 10 records (5 draft and 5 published) and you click on the Publish action - you can't select on the existing Published records, only the drafts. If it's decided that it is worth pursuing - I think this could be its own PR.

**Comment by ivan-kishko** [comment]
@lilidworkin @mxosman So I guess if behavior of records depending on bulk action is for another task then it is final commit. I made "new record" action only visible for admins, but bulk action dropdown was already only visible only for JC admins. I really do not remember did I do that or "select" action that was here before already used that condition. Anyway without contributor account I cannot test is so pls take a look and LMK if all good. Thanks!)

File: publisher/src/pages/Reports.tsx [hunk 6]
```diff
@@ -450,12 +454,14 @@ const Reports: React.FC = () => {
                           </Dropdown>
                         </BulkActionsDropdownContainer>
                       )}
-                      <ReportActionsButton
-                        textColor="blue"
-                        onClick={() => navigate("create")}
-                      >
-                        + New Record
-                      </ReportActionsButton>
+                      {isAdmin && (
```
Comment (mxosman) [inline publisher/src/pages/Reports.tsx:457]: I actually don't think this is necessary. If we look at [L412-413](https://github.com/Recidiviz/justice-counts/pull/488/commits/f242b46831010ea9a32615499949c9cffa69b115#diff-3a0d4bc45f47d1ec23abea296c6e2261928e9d1bf5655d982b8f2719f320443bR412-R413) - we already do this check and it's already within the scope of this conditional.

I think this meets all the criteria already without any changes.

JC Admin:
* Create Records button
* Bulk Action dropdown (Delete, Publish and Unpublish)

Agency Admin:
* Create Records button

Contributor:
* No Create Record button
* No Bulk Action dropdown

**Review by mxosman — APPROVED** [review]
One small note regarding the latest commit - otherwise, LGTM!

File: publisher/src/pages/Reports.tsx (hunk 6 above)
Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx:457]: yes you are right. didnt notice. Someone already did that admin stuff, gonna revert changes

Comment (ivan-kishko) [reply publisher/src/pages/Reports.tsx]: [4146ef7](https://github.com/Recidiviz/justice-counts/pull/488/commits/4146ef738603d32597ddecab8876f374d4e0509e)

**Review by lilidworkin — APPROVED** [review]
Awesome! Thanks @ivan-kishko and thank you for checking that @mxosman !

**Comment by hobuobi** [comment]
This is looking great! Thanks for tackling this. Only a few small comments: 

- If possible, the gray highlight on hover should extend to the edge of the screen. This might not have been a change you made, but it would be great to address it in this PR or a follow-up. 
Screen Shot 2023-03-13 at 9 52 33 PM
- You mentioned enabling or disabling records depending on the bulk action (e.g., disabling Published and Not Started records when you click Publish, or disabling Not Started and Draft when you click Unpublish). I'm fine with this going into another PR if that's what you decide to do, but we should not ship this feature to prod until that is implemented!

**Comment by ivan-kishko** [comment]
> This is looking great! Thanks for tackling this. Only a few small comments:
> 
> * If possible, the gray highlight on hover should extend to the edge of the screen. This might not have been a change you made, but it would be great to address it in this PR or a follow-up.
> 
> Screen Shot 2023-03-13 at 9 52 33 PM
> 
> - You mentioned enabling or disabling records depending on the bulk action (e.g., disabling Published and Not Started records when you click Publish, or disabling Not Started and Draft when you click Unpublish). I'm fine with this going into another PR if that's what you decide to do, but we should not ship this feature to prod until that is implemented!

Hello @hobuobi. Here it is [0753f96](https://github.com/Recidiviz/justice-counts/pull/488/commits/0753f96d21724b6c90f0d67e4e269eaa3b33e19e) (Row background color). And I guess @lilidworkin task should be made for selection behavior on "All Records" tab.

**Comment by lilidworkin** [comment]
Here you go @ivan-kishko ! https://github.com/Recidiviz/justice-counts/issues/498

Feel free to merge in this PR and close out the other ticket. Thanks!

**Pull request merged**